docs/Toy/toy/dialect.py

```diff
@@ -0,0 +1,267 @@
+"""
+Toy language dialect from MLIR tutorial.
+"""
+
+from __future__ import annotations
+
+from typing import Annotated, List, TypeAlias, Union, Optional, Any, cast
+
+from xdsl.ir import Dialect, SSAValue
+from xdsl.dialects.builtin import (Float64Type, FunctionType, Attribute,
+                                   FlatSymbolRefAttr, TensorType,
+                                   UnrankedTensorType, f64,
+                                   DenseIntOrFPElementsAttr, AnyTensorType,
+                                   StringAttr)
+from xdsl.irdl import (OpAttr, Operand, OptOpAttr, OptOperand, VarOpResult,
+                       VarOperand, irdl_op_definition, AnyAttr, Block, Region,
+                       Operation, OpResult)
+from xdsl.utils.exceptions import VerifyException
+
+TensorTypeF64: TypeAlias = TensorType[Float64Type]
+UnrankedTensorTypeF64: TypeAlias = UnrankedTensorType[Float64Type]
+AnyTensorTypeF64: TypeAlias = TensorTypeF64 | UnrankedTensorTypeF64
+
+
+@irdl_op_definition
+class ConstantOp(Operation):
+    """
+    Constant operation turns a literal into an SSA value. The data is attached
+    to the operation as an attribute. For example:
+
+    ```mlir
+      %0 = toy.constant dense<[[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]>
+                        : tensor<2x3xf64>
+    ```
+    """
+    name: str = "toy.constant"
+    value: OpAttr[DenseIntOrFPElementsAttr]
+    res: Annotated[OpResult, TensorTypeF64]
+
+    @staticmethod
+    def from_list(data: List[float], shape: List[int]):
+        value = DenseIntOrFPElementsAttr.tensor_from_list(data, f64, shape)
+
+        return ConstantOp.create(result_types=[value.type],
+                                 attributes={"value": value})
+
+    def verify_(self) -> None:
+        if not self.res.typ == self.value.type:
+            raise VerifyException(
+                "Expected value and result types to be equal: "
+                f"{self.res.typ}, {self.value.type}")
+
+
+@irdl_op_definition
+class AddOp(Operation):
+    """
+    The "add" operation performs element-wise addition between two tensors.
+    The shapes of the tensor operands are expected to match.
+    """
+    name: str = 'toy.add'
+    arguments: Annotated[VarOperand, AnyTensorTypeF64]
+    res: Annotated[OpResult, AnyTensorTypeF64]
+
+    @classmethod
+    def from_summands(cls: type[AddOp], lhs: SSAValue, rhs: SSAValue) -> AddOp:
+        assert isinstance(lhs.typ, TensorType | UnrankedTensorType)
+        element_type = cast(Float64Type,
+                            cast(TensorType[Any], lhs.typ).element_type)
+        return super().create(result_types=[element_type], operands=[lhs, rhs])
+
+
+@irdl_op_definition
+class FuncOp(Operation):
```

Comment:
What is the reason to use this and not the internal FuncOp? Just to show how it is defined?

Reply:
Yep. Do keep in mind that this is a reimplementation of the toy dialect in MLIR, and the goal is to be fully compatible.